doc: add dedicated machines private location documentation, closes HYB-411 (#178)
gatling/frontline-cloud-doc: content/reference/admin/private_locations/configuration/dedicated/index.md

---
title: "Dedicated Machines"
description: "Deploy Load Generators on dedicated machines that you manage."
lead: "Deploy Load Generators on dedicated machines that you manage."
date: 2023-01-12T16:46:04+00:00
lastmod: 2023-10-13T08:10:39+00:00
weight: 22055
---

> ## Dedicated Machines
> 
> To use Dedicated Machines private locations, the control plane must have access to your dedicated machines.
> 
> To set up a dedicated location configuration, ensure that each host is reachable through SSH on a specific port.
> The control plane possesses a private key for establishing connections, while the corresponding public key is shared and configured on every host in the location.

Reviewer: An introduction paragraph?

```suggestion
## Dedicated Machines

You can configure private locations with already existing servers. The control plane will be able to use configured dedicated machines as load generator during your simulations.

The control plane must have access to your dedicated machines. Ensure that each host is reachable through SSH on a specific port.
The control plane possesses a private key for establishing connections, while the corresponding public key is shared and configured on every host in the location.
```

- Java runtime environment
- `bash` the GNU Project's shell
- `jq` a lightweight and flexible command-line JSON processor
- `curl` command line tool for transferring data with URLs
- `~/.ssh/authorized_keys` with control plane public key

### Control plane configuration file

```bash
control-plane {
  # Control plane token
  token = "cpt_example_c7oze5djp3u14a5xqjanh..."
  # Control plane token with an environment variable
  token = ${?CONTROL_PLANE_TOKEN}
  # Control plane description (optional)
  description = "my control plane description"
  # Locations configurations
  locations = [
    {
      # Private location ID, must be prefixed by prl_, only consist of numbers 0-9, 
      # lowercase letters a-z, and underscores, with a max length of 30 characters
      id = "prl_private_location_example"
      # Private location description (optional)
      description = "Private Location on Dedicated Machines"
      # Private location type
      type = "dedicated"
      ssh {
        # SSH user
        user = "gatling"
        ## SSH private key
        private-key {
          # SSH private key path used to secure connections
          path = "private/key/path"

>           ## SSH private key password (optional, default: none)

Reviewer: ```suggestion
          # SSH private key password (optional, default: none)
```

          password = "password"
           # SSH private key password with an environment variable
          password = ${?PRIVATE_KEY_PASSWORD}
        }

>         # SSH port (optional, default: 22)
>         # port = 22
>         # SSH connection timeout (optional, default 15 seconds)
>         # connection-timeout = 15 seconds

Reviewer: Should we have a convention that comment on commented code should be commented? :thinking: 

```suggestion
        # # SSH port (optional, default: 22)
        # port = 22
        # # SSH connection timeout (optional, default 15 seconds)
        # connection-timeout = 15 seconds
```

or:

```suggestion
        ## SSH port (optional, default: 22)
        # port = 22
        ## SSH connection timeout (optional, default 15 seconds)
        # connection-timeout = 15 seconds
```

      }

>       # Working directory on hosts, must be executable (optional, default: /tmp)
>       # working-directory = "/tmp

Reviewer: ```suggestion
      ## Working directory on hosts, must be executable (optional, default: /tmp)
      # working-directory = "/tmp
```

      
      # Hosts accessible with the SSH private key, on configured port (hostnames or IP addresses)
      hosts = [
        "localhost",
        "0.0.0.0"
      ]
      

>       # Java configuration (following configuration properties are optional)
>       # System properties (optional)
>       system-properties {
>         # ExampleKey = ExampleValue
>       }

Reviewer: ```suggestion
      ## Java configuration (following configuration properties are optional)
      ## System properties (optional)
      # system-properties {
        ## ExampleKey = ExampleValue
      #}
```

>       # Overwrite JAVA_HOME definition (optional)
>       # java-home = "/usr/lib/jvm/zulu"
>       # JVM Options (optional)
>       # Default ones, that can be overriden with precedence:
>       # [
>       #   "-Xmx4G", 
>       #   "-XX:MaxInlineLevel=20", 
>       #   "-XX:MaxTrivialSize=12", 
>       #   "-XX:+IgnoreUnrecognizedVMOptions", 
>       #   "--add-opens=java.base/java.nio=ALL-UNNAMED", 
>       #   "--add-opens=java.base/jdk.internal.misc=ALL-UNNAMED"
>       # ]
>       # Based on your instance configuration, you may want to update Xmx and Xms values.
>       # jvm-options = ["-Xmx4G", "-Xms512M"]

Reviewer: ```suggestion
      ## Overwrite JAVA_HOME definition (optional)
      # java-home = "/usr/lib/jvm/zulu"
      ## JVM Options (optional)
      ## Default ones, that can be overriden with precedence:
      ## [
      ##   "-Xmx4G", 
      ##   "-XX:MaxInlineLevel=20", 
      ##   "-XX:MaxTrivialSize=12", 
      ##   "-XX:+IgnoreUnrecognizedVMOptions", 
      ##   "--add-opens=java.base/java.nio=ALL-UNNAMED", 
      ##   "--add-opens=java.base/jdk.internal.misc=ALL-UNNAMED"
      ## ]
      ## Based on your instance configuration, you may want to update Xmx and Xms values.
      # jvm-options = ["-Xmx4G", "-Xms512M"]
```

But as the default values are the one commented by default...

```suggestion
      ## Overwrite JAVA_HOME definition (optional)
      # java-home = "/usr/lib/jvm/zulu"

      ## JVM Options (optional)
      ## Based on your instance configuration, you may want to update Xmx and Xms values.
      # jvm-options = [
      #   "-Xmx4G", 
      #   "-XX:MaxInlineLevel=20", 
      #   "-XX:MaxTrivialSize=12", 
      #   "-XX:+IgnoreUnrecognizedVMOptions", 
      #   "--add-opens=java.base/java.nio=ALL-UNNAMED", 
      #   "--add-opens=java.base/jdk.internal.misc=ALL-UNNAMED"
      # ]
```

Author: Maybe one day, but not today (must update all related documentations)

    }
  ]
}
```

{{< alert warning >}}
Working directory must be executable
{{< /alert >}}

## Troubleshooting

### SSH keys

Execute the command `ssh-keygen -t ed25519 -f control-plane` to generate your key pair. _(You have the option to configure a password during this process)_

This will generate the `control-plane` private key and `control-plane.pub` public key in the current directory where the command was executed.

Next, you'll need to configure the authorized keys on the host for a user named, for the example, `gatling`.

To do this, follow these steps:

> 1. Create the authorized_keys file in the gatling/.ssh directory on the host:
> ```bash
> touch gatling/.ssh/authorized_keys
> chmod 600 gatling/.ssh/authorized_keys
> ```

Reviewer: I guess the code block is part of the item in the list.

````suggestion
1. Create the authorized_keys file in the gatling/.ssh directory on the host:
    ```bash
    touch gatling/.ssh/authorized_keys
    chmod 600 gatling/.ssh/authorized_keys
    ```
````

> 2. Add the public key to gatling/.ssh/authorized_keys. You can do this either by using the ssh-copy-id command:

Reviewer 2: ```suggestion
2. Add the public key to gatling/.ssh/authorized_keys on the host. You can do this either by using the ssh-copy-id command:
```

> ```bash
> ssh-copy-id -i control-plane.pub gatling@host
> ```
> Or, manually by appending the public key to the file with:
> ```bash
> cat control-plane.pub >> gatling/.ssh/authorized_keys
> ```

Reviewer: ````suggestion
    ```bash
    ssh-copy-id -i control-plane.pub gatling@host
    ```

    Or, manually by appending the public key to the file with:

    ```bash
    cat control-plane.pub >> gatling/.ssh/authorized_keys
    ```
````

> 3. Add the private key to the control plane

Reviewer 2: ```suggestion
3. Add the private key to the control plane. Refer to your control plane config file for the expected private key file path: for your location, the value of `ssh.private-key.path`.
```

### Timeout

If your execution results in a timeout without any additional information, it indicates that the script failed unexpectedly. 
To investigate this issue, you can configure a location to remain active by using the following settings:

```
locations = [
  {
    id = "prl_dedicated"
    type = "dedicated"
    ...
    debug {
      keep-load-generator-alive = true
    }
  }
]
```

> Afterward, you can examine the logs via systemd logs using the identifier 'gatling' with `journalctl --identifier gatling`.

Reviewer: ```suggestion
Afterward, you can examine the logs via systemd logs using the identifier '`gatling`' with `journalctl --identifier gatling`.
```

---
title: "Dedicated Machines"
description: "Deploy Load Generators on dedicated machines that you manage."
lead: "Deploy Load Generators on dedicated machines that you manage."
date: 2023-01-12T16:46:04+00:00
lastmod: 2023-10-13T08:10:39+00:00
weight: 22055
---

## Dedicated Machines

> You can configure private locations with already existing servers. 

Reviewer 3: ```suggestion
You can configure private locations with pre-existing servers. 
```

> The control plane will be able to use configured dedicated machines as load generator during your simulations.

Reviewer 3: ```suggestion
The control plane will be able to use configured dedicated machines as load generators during your simulations.
```

The control plane must have access to your dedicated machines. Ensure that each host is reachable through SSH on a specific port.

The control plane possesses a private key for establishing connections, while the corresponding public key is shared and configured on every host in the location.

### Host prerequisites

> Location hosts environment must include:

Reviewer 3: ```suggestion
The location host environment must include:
```

- Java runtime environment
- `bash` the GNU Project's shell

> - `jq` a lightweight and flexible command-line JSON processor

Reviewer 3: Should we provide external links to any of these?

- `curl` command line tool for transferring data with URLs

> - `~/.ssh/authorized_keys` with control plane public key

Reviewer 3: ```suggestion
- `~/.ssh/authorized_keys` with a control plane public key
```

      }
      # Working directory on hosts, must be executable (optional, default: /tmp)
      # working-directory = "/tmp
      
      # Hosts accessible with the SSH private key, on configured port (hostnames or IP addresses)
      hosts = [
        "localhost",
        "0.0.0.0"
      ]
      
      # Java configuration (following configuration properties are optional)
      # System properties (optional)
      system-properties {
        # ExampleKey = ExampleValue
      }
      # Overwrite JAVA_HOME definition (optional)
      # java-home = "/usr/lib/jvm/zulu"
      # JVM Options (optional)
      # Default ones, that can be overriden with precedence:
      # [
      #   "-Xmx4G", 
      #   "-XX:MaxInlineLevel=20", 
      #   "-XX:MaxTrivialSize=12", 
      #   "-XX:+IgnoreUnrecognizedVMOptions", 
      #   "--add-opens=java.base/java.nio=ALL-UNNAMED", 
      #   "--add-opens=java.base/jdk.internal.misc=ALL-UNNAMED"
      # ]
      # Based on your instance configuration, you may want to update Xmx and Xms values.
      # jvm-options = ["-Xmx4G", "-Xms512M"]
    }
  ]
}
```

{{< alert warning >}}
Working directory must be executable
{{< /alert >}}

## Troubleshooting

> 

Reviewer 3: ```suggestion
This section contains tips for troubleshooting common issues with dedicated machines. If you have an issue not covered here, consider posting a question on the[ Community Forum](https://community.gatling.io/c/enterprise/6). 
```

### SSH keys

Execute the command `ssh-keygen -t ed25519 -f control-plane` to generate your key pair. _(You have the option to configure a password during this process)_

This will generate the `control-plane` private key and `control-plane.pub` public key in the current directory where the command was executed.

> Next, you'll need to configure the authorized keys on the host for a user named, for the example, `gatling`.

Reviewer 3: ```suggestion
Next, you'll need to configure the authorized keys on the host for a user named, for example, `gatling`.
```

Reviewer 3: I was a little confused by this, but I think removing 'the' fixes it.

> To do this, follow these steps:

Reviewer 3: ```suggestion
To configure the authorized keys on the host:
```

Reviewer 3: Here you can omit the 'follow these steps' but explicitly say what it is for.

1. Create the authorized_keys file in the gatling/.ssh directory on the host:
  ```bash
  touch gatling/.ssh/authorized_keys
  chmod 600 gatling/.ssh/authorized_keys
  ```

2. Add the public key to gatling/.ssh/authorized_keys on the host. You can do this either by using the ssh-copy-id command:
  ```bash
  ssh-copy-id -i control-plane.pub gatling@host
  ```
  Or, manually by appending the public key to the file with:
  ```bash
  cat control-plane.pub >> gatling/.ssh/authorized_keys
  ```

3. Add the private key to the control plane. Refer to your control plane config file for the expected private key file path: for your location, the value of `ssh.private-key.path`.

> 

Reviewer 3: This is a perfectly written procedure!😀
